### ISSUE TITLE
"FLOAT" alias for float numbers does not produce an error despite not being supported by BigQuery

### ISSUE THREAD

**Issue by VMois** [sqlfluff/sqlfluff#6772]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

In the BQ dialect, the only accepted alias for float number is `FLOAT64` - https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#floating_point_types

But linter for example will not throw any errors (but it will not run in BQ):

```sql
SELECT CAST("4.5" AS FLOAT)
```

### Expected Behaviour

The linter should mark it as a syntax issue or similar.

### Observed Behaviour

I would expect linter to fail here as it is not a valid BQ syntax. Not sure if it is a dialect parsing issue or maybe there should be a dialect-specific rule to flag. I lean more towards that BQ dialect should define more strictly allowed data types.

From code perspective, looks like BQ dialect relies on [DatatypeSegment](https://github.com/sqlfluff/sqlfluff/blob/2004faf0145f7c894bac6994a88ff0faa499aa7b/src/sqlfluff/dialects/dialect_bigquery.py#L1277) to define types which in terms relies on generic [DatatypeIdentifierSegment](https://github.com/sqlfluff/sqlfluff/blob/2004faf0145f7c894bac6994a88ff0faa499aa7b/src/sqlfluff/dialects/dialect_ansi.py#L340). I can see a few comments about more explicit and restrictive datatypes.

I would be interested in looking at how we can define stricter datatypes for BigQuery. Happy to submit MR. I would also appreciate any guidelines and resources from the core team regarding the issue. Maybe there were bigger concerns in the past to leave types for BQ open. 

Thank you.

### How to reproduce

Linter has no errors, it fails in BQ
```sql
SELECT CAST("4.5" AS FLOAT)
SELECT CAST("4.5" AS FLOAT2)  # as you can see any alias works
```

Linter has no errors, it works in BQ

```sql
SELECT CAST("4.5" AS FLOAT64)
```

### Dialect

BigQuery

### Version

sqlfluff, version 3.3.1

### Configuration

default one

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by moreaupascal56** [comment]
Hi @VMois,

I would like to dig around this as well. As you said It seems to me that the behavior for now is that the data types just cannot be from the reserved list of keywords by dialect but it can be anything else.

cf. 
https://github.com/sqlfluff/sqlfluff/blob/d3a3ff832cbe2d3cd33d4fc41e7238b7ef43992d/src/sqlfluff/dialects/dialect_ansi.py#L340-L352


It is done in the dialect_ansi so, I guess there is no type enforcement in any dialects (but maybe there is it should be confirmed, if there is we could duplicate the behavior). You said it is only BigQuery but I think at least in Snowflake it seems to be the same thing.

I would be happy to help on this if needed.

**Comment by VMois** [comment]
You are right probably - all dialects are affected. I am mostly just interested in BigQuery one, that is why I provided details there.

I was waiting for the main contributors response on that to confirm if this is something they would like to see added. Next few weeks I will have more time and might try to dig into this deeper. If you want to look into this as well, please, go ahead.

I believe we will need to "onboard" each dialect separately but solution can be a generic one (e.g. custom datatype for each dialect).
